patch/llpcPatchResourceCollect.cpp

```diff
@@ -3501,6 +3506,250 @@ void PatchResourceCollect::ReassembleOutputExportCalls()
+        }
+        if (auto pShuffle = dyn_cast<ShuffleVectorInst>(pUser))
+        {
+            SmallVector<int, 4> mask;
+            pShuffle->getShuffleMask(mask);
+            for (int maskElement : mask)
+            {
+                if (maskElement >= 0)
+                {
+                    if (maskElement < scalarizeBy)
+                    {
+                        if (pShuffle->getOperand(0) == pCall)
+                        {
+                            elementUsed[maskElement] = true;
+                        }
+                    }
+                    else
+                    {
+                        LLPC_ASSERT(maskElement < 2 * scalarizeBy);
+                        if (pShuffle->getOperand(1) == pCall)
+                        {
+                            elementUsed[maskElement - scalarizeBy] = true;
+                        }
+                    }
+                }
+            }
+            continue;
+        }
+        unknownElementsUsed = true;
+        break;
+    }
+
+    // Load the individual elements and insert into a vector.
+    Value* pResult = UndefValue::get(pResultTy);
+    std::string callName = isInterpolant ? LlpcName::InputImportInterpolant : LlpcName::InputImportGeneric;
+    AddTypeMangling(pElementTy, args, callName);
+    for (uint32_t i = 0; i != scalarizeBy; ++i)
+    {
+        if (!unknownElementsUsed && !elementUsed[i])
+        {
+            continue; // Omit trivially unused element
+        }
+        args[elemIdxArgNum] = builder.getInt32(elemIdx + i);
+
+        CallInst* pElement = EmitCall(callName, pElementTy, args, Attribute::ReadOnly, builder);
+        pResult = builder.CreateInsertElement(pResult, pElement, i);
+        if (pElementTy->getPrimitiveSizeInBits() == 64)
+        {
+            // If scalarizing with 64 bit elements, further split each element.
+            ScalarizeInputGeneric(pElement);
+        }
+    }
+
+    pCall->replaceAllUsesWith(pResult);
+    pCall->eraseFromParent();
+}
+
+// =====================================================================================================================
+// Scalarize a generic output.
+// This is known to be a final vertex stage (VS/TES/GS) generic output that is either a vector or 64 bit.
+void PatchResourceCollect::ScalarizeOutputGeneric(
+    CallInst* pCall)  // [in] The llpc.output.export.generic call
+{
+    IRBuilder<> builder(pCall->getContext());
+    builder.SetInsertPoint(pCall);
+
+    // VS:  @llpc.output.export.generic.%Type%(i32 location, i32 elemIdx, %Type% outputValue)
+    // TES: @llpc.output.export.generic.%Type%(i32 location, i32 elemIdx, %Type% outputValue)
+    // GS:  @llpc.output.export.generic.%Type%(i32 location, i32 elemIdx, i32 streamId, %Type% outputValue)
+    SmallVector<Value*, 5> args;
+    for (uint32_t i = 0, e = pCall->getNumArgOperands(); i != e; ++i)
+    {
+        args.push_back(pCall->getArgOperand(i));
+    }
+
+    static const uint32_t ElemIdxArgNum = 1;
```

Comment:
ElemIdxArgNum  -> ArgIdxOfElemIdx
In LLPC, because num could denote both the total count and the ordinal, so we avoid using it. If we want the total number, use "XxxCount", if we want the ordinal, name it "XxxIdx"

Reply:
I'll use ElemIdxArgIdx to avoid the awkward "AOfB" construct.